File: src/report.ts
/**
 * Conformance report types — GC-2-C-01.
 *
 * Defines the machine-readable output contract for all Governance Commons validators.
 * Every validator MUST produce a ConformanceReport whose serialised form validates against:
 *   docs/conformance/conformance-report.schema.json  (gc_report_version: "1.0.0")
 */

/** Version of the GC conformance report format schema. */
export const GC_REPORT_VERSION = "1.0.0" as const;

/** Conformance profile evaluated by the validator. */
export type ConformanceProfile = "advisory" | "standard" | "strict";

/** Outcome of evaluating a single rule. */
export type RuleOutcome = "pass" | "fail" | "skip";

/** Highest conformance level achieved by the subject. */
export type ConformanceLevel = "none" | "advisory" | "standard" | "strict";

/** Per-rule evaluation result. */
export interface RuleResult {
  readonly rule_id: string;
  readonly description: string;
  readonly result: RuleOutcome;
  readonly message?: string | null;
}

/** Aggregate counts across all rule results. */
export interface ReportSummary {
  readonly total: number;
  readonly passed: number;
  readonly failed: number;
  readonly skipped: number;
}

/** Machine-readable conformance report. Serialises to the GC conformance report schema v1.0.0. */
export interface ConformanceReport {
  readonly gc_report_version: typeof GC_REPORT_VERSION;
  readonly spec: string;
  readonly spec_version: string;
  readonly profile: ConformanceProfile;
  readonly subject: string;
  readonly generated_at: string;
  readonly rules: readonly RuleResult[];
  readonly summary: ReportSummary;
  readonly conformant: boolean;
  readonly conformance_level: ConformanceLevel;
}

/** Options for {@link buildReport}. */
export interface BuildReportOptions {
  readonly spec: string;
  readonly spec_version: string;
  readonly profile: ConformanceProfile;
  readonly subject: string;
  readonly rules: readonly RuleResult[];
  /** ISO 8601 UTC timestamp; defaults to `new Date().toISOString()`. */
  readonly generated_at?: string;
}

/**
 * Construct a ConformanceReport from a list of RuleResults.
 *
 * Computes summary counts, conformant flag, and conformance_level automatically.
 *
 * @example
 * const report = buildReport({
 *   spec: "ons",
 *   spec_version: "1.4.0",
 *   profile: "standard",
 *   subject: "my-file.yaml",
 *   rules: [{ rule_id: "ONS-CASING-001", description: "snake_case identifiers", result: "pass" }],
 * });
 */
export function buildReport(options: BuildReportOptions): ConformanceReport {
  const { spec, spec_version, profile, subject, rules } = options;
  const generated_at = options.generated_at ?? new Date().toISOString().replace(/\.\d{3}Z$/, "Z");

  const passed = rules.filter((r) => r.result === "pass").length;
  const failed = rules.filter((r) => r.result === "fail").length;
  const skipped = rules.filter((r) => r.result === "skip").length;
  const conformant = failed === 0;
  const conformance_level: ConformanceLevel = conformant ? profile : "none";

  return {
    gc_report_version: GC_REPORT_VERSION,
    spec,
    spec_version,
    profile,
    subject,
    generated_at,
    rules,
    summary: { total: passed + failed + skipped, passed, failed, skipped },
    conformant,
    conformance_level,
  };
}

File: governance_commons/report.py
"""
Conformance report types — GC-2-C-01.

Defines the machine-readable output contract for all Governance Commons validators.
Every validator (Python or TypeScript) MUST produce a ConformanceReport whose
serialised form validates against:
  docs/conformance/conformance-report.schema.json  (gc_report_version: "1.0.0")
"""

from __future__ import annotations

import json
from dataclasses import asdict, dataclass, field
from datetime import datetime, timezone
from typing import Literal

GC_REPORT_VERSION = "1.0.0"

ConformanceProfile = Literal["advisory", "standard", "strict"]
RuleOutcome = Literal["pass", "fail", "skip"]
ConformanceLevel = Literal["none", "advisory", "standard", "strict"]


@dataclass
class RuleResult:
    rule_id: str
    description: str
    result: RuleOutcome
    message: str | None = None


@dataclass
class ReportSummary:
    total: int
    passed: int
    failed: int
    skipped: int


@dataclass
class ConformanceReport:
    spec: str
    spec_version: str
    profile: ConformanceProfile
    subject: str
    generated_at: str
    rules: list[RuleResult]
    summary: ReportSummary
    conformant: bool
    conformance_level: ConformanceLevel
    gc_report_version: str = field(default=GC_REPORT_VERSION, init=False)

    def to_dict(self) -> dict:
        return asdict(self)

    def to_json(self, *, indent: int | None = 2) -> str:
        return json.dumps(self.to_dict(), indent=indent)


def build_report(
    *,
    spec: str,
    spec_version: str,
    profile: ConformanceProfile,
    subject: str,
    rules: list[RuleResult],
    generated_at: str | None = None,
) -> ConformanceReport:
    """Construct a ConformanceReport from a list of RuleResults.

    Computes summary counts, conformant flag, and conformance_level automatically.

    Args:
        spec: Spec identifier (e.g. "ons").
        spec_version: Semver of the spec (e.g. "1.4.0").
        profile: Conformance profile evaluated.
        subject: File path, URI, or label of what was validated.
        rules: Ordered list of per-rule results.
        generated_at: ISO 8601 UTC timestamp; defaults to now.

    Example::

        report = build_report(
            spec="ons",
            spec_version="1.4.0",
            profile="standard",
            subject="my-file.yaml",
            rules=[RuleResult("ONS-CASING-001", "snake_case identifiers", "pass")],
        )
    """
    passed = sum(1 for r in rules if r.result == "pass")
    failed = sum(1 for r in rules if r.result == "fail")
    skipped = sum(1 for r in rules if r.result == "skip")
    conformant = failed == 0
    conformance_level: ConformanceLevel = profile if conformant else "none"

    return ConformanceReport(
        spec=spec,
        spec_version=spec_version,
        profile=profile,
        subject=subject,
        generated_at=generated_at or datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ"),
        rules=rules,
        summary=ReportSummary(total=passed + failed + skipped, passed=passed, failed=failed, skipped=skipped),
        conformant=conformant,
        conformance_level=conformance_level,
    )
